File: homework_prompts/rb/w3_d2_quasilinear_sorts.rb
#  Homework 08 - Quasilinear Sorts
#
#  Problem 1: Mergesort
#
#  Prompt:    Given an unsorted array of integers, return the array
#             sorted using mergesort.
#
#             What are the time and auxilliary space complexity?
#
#  Input:     An unsorted array of integers
#  Output:    A sorted array of integers
#
#  Example:   input = [3,9,1,4,7] , output = [1,3,4,7,9]
#

# Time Complexity:
# Auxiliary Space Complexity:
def mergesort(input)
end


############################################################
###############  DO NOT TOUCH TEST BELOW!!!  ###############
############################################################

# custom expect function to handle tests
# List count : keeps track out how many tests pass and how many total
#   in the form of a two item array i.e., [0, 0]
# String name : describes the test
# Function test : performs a set of operations and returns a boolean
#   indicating if test passed
def expect(count, name, test)
  count[1] = count[1] + 1

  result = 'false'
  errMsg = nil
  begin
    if test()
      result = ' true'
      count[0] = count[0] + 1
    end
  rescue NameError => e
    MyModule::Logger.error("Error loading the deployer #{class_name}. This deployer it's not installed!")
  rescue StandardError => err
    errMsg = err.message
  end

  puts'  ' + (count[1]).to_s + ')   ' + result + ' : ' + name
  if errMsg != nil
    puts'       ' + errMsg + '\n'
  end
end

# code for capturing puts output
require 'stringio'
require 'ostruct'

class Capture

  def self.capture &block

    # redirect output to StringIO objects
    stdout = StringIO.new
    $stdout = stdout

    result = block.call

    # restore normal output
    $stdout = STDOUT

    OpenStruct.new result: result, stdout: stdout.string
  end
end

# code for checking if lists are equal
def arrays_equal(arr1, arr2)
  if arr1.length != arr2.length
    return false
  end
  for i in 0..arr1.length-1
    if arr1[i] != arr2[i]
      return false
    end
  end
  return true
end

# code for checking if array is sorted (linear runtime)
def is_sorted(input)
  if (input.length < 2)
    return true
  end
  for i in 1..input.length - 1
    if (input[i-1] > input[i])
      return false
    end
  end
  return true
end



puts 'Merge Sort Tests'
test_count = [0, 0]

def test()
  example = mergesort([3,9,1,4,7])
  return arrays_equal(example, [1,3,4,7,9])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should sort example input', test)

def test()
  example = mergesort([])
  return arrays_equal(example, [])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return empty array for empty input', test)

def test()
  example = mergesort([10])
  return arrays_equal(example, [10])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should sort single-element input', test)

def test()
  work = []
  for i in 0..999
    work.push(Random.rand(1000))
  end
  example = mergesort(work)
  return example.length == 1000 && is_sorted(example)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should sort moderate-sized input', test)

def test()
  work = []
  for i in 0..999999
    work.push(Random.rand(1000000))
  end
  example = mergesort(work)
  return example.length == 1000000 && is_sorted(example)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should sort large input', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"

File: homework_prompts/swift/w3_d2_quasilinear_sorts.swift
/*
*  Homework 08 - Quasilinear Sorts
*
*  Problem 1: Mergesort
*
*  Prompt:    Given an unsorted array of integers, return the array
*             sorted using mergesort.
*
*             What are the time and auxilliary space complexity?
*
*  Input:     An unsorted array of integers
*  Output:    A sorted array of integers
*
*  Example:   input = [3,9,1,4,7] , output = [1,3,4,7,9]
*
*/

import Darwin


// Worse Time Complexity:
// Worse Auxiliary Space Complexity:
// Average Time Complexity:
// Average Auxiliary Space Complexity:
func mergesort(_ input: inout [Int]) -> [Int] {
  // YOUR WORK HERE
  return [Int]()
}


/**
 *  Problem 2: Bucket Sort
 *
 *  Prompt:    Given an unsorted array of numbers which are randomly distributed
 *             across the a range, sort the numbers using bucket sort.
 *
 *  Input:     arr: [Float]
 *  Input:     {Float} - minimum of range (inclusive)
 *  Input:     {Float} - maximum of range (exclusive)
 *
 *  Output:    {Array}
 *
 *  Example:   [0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434], 0.0, 1.0 -->
 *             [0.1234, 0.3434, 0.565, 0.656, 0.665, 0.897]
 *
 *  HINT:      Use either N buckets or square root of N buckets
 *             Recursively sort each bucket using bucketsort until the bucket
 *             size is less than 50.
 */


 // Time Complexity:
 // Auxiliary Space Complexity:
 func bucketsort(_ arr: [Float], _ min: Float, _ max: Float) -> [Float] {
   // YOUR WORK HERE
   return [Float]()
 }


 /*
  *
  *  Problem 3: Kth Smallest Element in a Range
  *
  *  Prompt:    Given an unsorted array of whole integers in the range
  *             1000 to 9000, find the Kth smallest element in linear time
  *             The array can have duplicates.
  *
  *  Input:     Unsorted array of whole integers in range of 1000 to 9000
  *             Kth smallest element you want to find
  *
  *  Output:    Kth smalest element in the range
  *
  *  Example:   array = [1984, 1337, 9000, 8304, 5150, 9000, 8304], k=5
  *             output = 8304
  *
  *  Notes:     What are the time and auxilliary space complexity?
  *
  */

 // Time Complexity:
 // Auxiliary Space Complexity:
 func kthSmallest(_ arr: [Int], _ k: Int) -> Int{
   // YOUR WORK HERE
   return -1
 }




////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}

func arraysEqual(_ arr1: [Int], _ arr2: [Int]) -> Bool {
  if arr1.count != arr2.count {
    return false
  }
  for i in 0..<arr1.count {
    if arr1[i] != arr2[i] {
      return false
    }
  }
  return true
}


print("Mergesort Sort Tests")
var testCount: [Int] = [0,0]

assert(&testCount, "should sort [3,9,1,4,7]",  {
  var arr = [3,9,1,4,7]
  let test = mergesort(&arr)
  return arraysEqual(test, [1,3,4,7,9])
})

assert(&testCount, "should return empty array for empty input",  {
  var arr = [Int]()
  let test = mergesort(&arr)
  return arraysEqual(test, [])
})

assert(&testCount, "should sort single-element input",  {
  var arr = [10]
  let test = mergesort(&arr)
  return arraysEqual(test, [10])
})

assert(&testCount, "should sort moderate-sized input",  {
  var arr = [Int]()
  for _ in 1..<1000 {
    arr.append(Int(arc4random_uniform(1000)))
  }
  var arr2 = arr
  let test = mergesort(&arr2)
  return arraysEqual(test, arr.sorted())
})

assert(&testCount, "should sort large input",  {
  var arr = [Int]()
  for _ in 1..<1000000 {
    arr.append(Int(arc4random_uniform(1000000)))
  }
  var arr2 = arr
  let test = mergesort(&arr2)
  return arraysEqual(test, arr.sorted())
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n");

func arraysEqual(_ arr1: [Float], _ arr2: [Float]) -> Bool {
  if arr1.count != arr2.count {
    return false
  }
  for i in 0..<arr1.count {
    if arr1[i] != arr2[i] {
      return false
    }
  }
  return true
}

print("Bucketsort Tests")
testCount = [0,0]

assert(&testCount, "should sort example input",  {
  let example = bucketsort([0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434], 0.0, 1.0)
  return arraysEqual(example, [0.1234, 0.3434, 0.565, 0.656, 0.665, 0.897])
})

assert(&testCount, "should return empty array for empty input",  {
  let example = bucketsort([Float](), 0.0, 1.0)
  return arraysEqual(example, [Float]())
})

assert(&testCount, "should sort single-element input",  {
  let example = bucketsort([0.5], 0.0, 1.0)
  return arraysEqual(example, [0.5])
})

assert(&testCount, "should sort moderate-sized input",  {
  var arr = [Float]()
  for _ in 1..<1000 {
    arr.append(Float(arc4random_uniform(1000)) / 1000)
  }
  let arr2 = arr
  let test = bucketsort(arr2, 0.0, 1.0)
  return arraysEqual(test, arr.sorted())
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Kth Smallest Element Tests")
testCount = [0,0]

assert(&testCount, "should return 8304 for sample input",  {
  let example = kthSmallest([1984, 1337, 9000, 8304, 5150, 9000, 8304], 5)
  return example == 8304
})

assert(&testCount, "should return 1337 for 1st smallest element with sample input array",  {
  let example = kthSmallest([1984, 1337, 9000, 8304, 5150, 9000, 8304], 1)
  return example == 1337
})

assert(&testCount, "should error out when asking for kth smallest when k exceeds size of input array",  {
  let example = kthSmallest([1984, 1337, 9000, 8304, 5150, 9000, 8304], 10)
  return example == -1
})

assert(&testCount, "should work for single-element array",  {
  let example = kthSmallest([8304], 1)
  return example == 8304
})

assert(&testCount, "should work for a large array",  {
  var testCase = [Int]()
  for _ in 1..<1000000 {
    testCase.append(Int(arc4random_uniform(8000)) + 1000)
  }
  let example = kthSmallest(testCase, 21875)
  testCase = testCase.sorted()
  return example == testCase[21874]
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
